packages/flutter/lib/src/material_animated_icons/animated_icons.dart

```diff
@@ -0,0 +1,269 @@
+  final AnimatedIconData icon;
+
+  /// Semantic label for the icon.
+  ///
+  /// This would be read out in accessibility modes (e.g TalkBack/VoiceOver).
+  /// This label does not show in the UI.
+  ///
+  /// See also:
+  ///
+  ///  * [Semantics.label], which is set to [semanticLabel] in the underlying
+  ///    [Semantics] widget.
+  final String semanticLabel;
+
+  /// The text direction to use for rendering the icon.
+  ///
+  /// If this is null, the ambient [Directionality] is used instead.
+  ///
+  /// If the text diection is [TextDirection.rtl], the icon will be mirrored
+  /// horizontally (e.g back arrow will point right).
+  final TextDirection textDirection;
+
+  static final _UiPathFactory _pathFactory = () => new ui.Path();
+
+  @override
+  Widget build(BuildContext context) {
+    final _AnimatedIconData iconData = icon;
+    final IconThemeData iconTheme = IconTheme.of(context);
+    final double iconSize = size ?? iconTheme.size;
+    final TextDirection textDirection = this.textDirection ?? Directionality.of(context);
+    return new Semantics(
+      label: semanticLabel,
+      child: new CustomPaint(
+        size: new Size(iconSize, iconSize),
+        painter: new _AnimatedIconPainter(
+          iconData.paths,
+          progress,
+          color ?? iconTheme.color,
+          iconSize / iconData.size.bottomRight(const Offset(0.0, 0.0)).dx,
+          textDirection == TextDirection.rtl,
+          _pathFactory,
+        ),
+      ),
+    );
+  }
+}
+
+typedef ui.Path _UiPathFactory();
+
+class _AnimatedIconPainter extends CustomPainter {
+  _AnimatedIconPainter(
+    this.paths,
+    this.progress,
+    this.color,
+    this.scale,
+    this.shouldMirror,
+    this.uiPathFactory,
+  ) : super(repaint: progress);
+
+  // This list is assumed to be immutable, changes to the contents of the list
+  // will not trigger a redraw as shouldRepaint will keep returning false.
+  final List<_PathFrames> paths;
+  final Animation<double> progress;
+  final Color color;
+  final double scale;
+  /// If this is true the image will be mirrored horizontally.
+  final bool shouldMirror;
+  final _UiPathFactory uiPathFactory;
+
+  @override
+  void paint(ui.Canvas canvas, Size size) {
+    canvas.scale(scale, scale);
+    if (shouldMirror) {
+      canvas.rotate(math.pi);
+      canvas.translate(-size.width, -size.height);
+    }
+
+    for (_PathFrames path in paths)
+      path.paint(canvas, color, uiPathFactory, progress.value.clamp(0.0, 1.0));
+  }
+
+
+  @override
+  bool shouldRepaint(_AnimatedIconPainter oldDelegate) {
+    return oldDelegate.progress.value != progress.value
+      || oldDelegate.color != color
+      // We are comparing the paths list by reference, assuming the list is
+      // treated as immutable to be more efficient.
+      || oldDelegate.paths != paths
+      || oldDelegate.scale != scale
+      || oldDelegate.uiPathFactory != uiPathFactory;
```

Comment:
indenting is off; it should align the expressions vertically

Reply:
done